src/GafferImage/Resample.cpp

```diff
@@ -318,6 +345,626 @@ Box2f transform( const Box2f &b, const M33f &m )
+				// then add one open segment.
+				numVolumeSegmentsOpen += sampleType >= DepthSampleEnd;
+
+				if( sampleType >= DepthSampleInterpolated )
+				{
+					// Leverage the special way we store interpolated types - the value of the type is actually
+					// the index of the End sample for this segment, which is the next depth this source
+					// requires us to output.
+					sourceNextMandatory[i] = curSamples[ sampleType ].depth;
+				}
+				else
+				{
+					// If this isn't an interpolated sample, then we need to output this depth.
+					sourceNextMandatory[i] = curSamples[ nextIndex ].depth;
+				}
+
+				// Update the sourceNextSample value used to quick skip this source when it doesn't
+				// have anything to contribute.
+				sourceNextSample[i] = curSamples[ nextIndex ].depth;
+			}
+			else
+			{
+				// This source is done, we don't need to do more with it
+				sourceNextMandatory[i] = std::numeric_limits<float>::infinity();
+				sourceNextSample[i] = std::numeric_limits<float>::infinity();
+			}
+		}
+
+		// We've processed all the sources at this depth, now we can actually output a segment
+		if( !onlyStarts )
+		{
+			// For volume segments, the segment starts at the last depth we processed, for point samples,
+			// it starts right her.
+			if( !segmentIsVolume )
+			{
+				prevDepth = outputDepth;
+			}
+
+			if( totalAccumAlpha > 1.0f )
+			{
+				// We don't want floating point error to result in this going over 1.0 and making everything wonky
+				totalAccumAlpha = 1.0f;
+			}
+
+			// The totalAccumAlpha is accumulated additively. We need to convert that into a multiplicative
+			// alpha that will composite properly.
+			float segmentAlpha = ( totalAccumAlpha - prevTotalAccumAlpha ) / ( 1 - prevTotalAccumAlpha );
+
+			// Add a value to each of our hardcode output channels
+			outputAlpha.push_back( segmentAlpha );
+			outputZ.push_back( prevDepth );
+			outputZBack.push_back( outputDepth );
+
+			// And if we need to process arbitrary channels, we need to store how many contributions we
+			// need for this output
+			if( contributionCounts )
+			{
+				contributionCounts->push_back( contributionCount );
+			}
+
+			numSegments++;
+
+			prevTotalAccumAlpha = totalAccumAlpha;
+		}
+
+		// Ready for next loop
+		prevDepth = outputDepth;
+
+		if( !( totalSamplesDone > prevTotalSamplesDone ) )
+		{
+			// In theory, this should just be an assert, but if there is any logic error here
+			// triggered by a weird floating point value or something, this could lead to a hang.
+			// We definitely don't want to hang - if we complete an iteration without progress, throw
+			// an exception instead.
+			throw IECore::Exception(
+				"Internal failure in Resample while processing deep image. This should "
+				"not be possible. We would appreciate if you can isolate a crop of the image which triggers "
+				"the issue and submit it to the Gaffer dev team."
+			);
```

Comment:
This turned out to be quite a simple thing ... I'd copied the code for filterWeights from filterWeights1D, and expected it to match filterWeights2D, but there's what is probably a typo in there - a constant that should be "0.5f" is just "0.5", causing the calculation to happen in double precision, and get a non-matching result do to precision.

The other thing that came up when playing with this is that performance of resizing to a very small result is incredibly bad. I probably shouldn't worry about this for now, because it's not specific to how we handle deep: Resample matches a common convention where we don't worry about the dataWindow while processing tiles. This means that due to the tile size, resizing a 1000 pixel image down to 1 pixel is treated identically to resizing a 128000 pixel image down to 128 pixels. This is actually kinda less of a problem for deep images with boundingMode set to Black, because the deep processing is so much slower on pixels that aren't empty ( whereas using a non-seperable filter to resize a 2D image down to one pixel would be extremely awful for no reason ), but if boundingMode is set to Clamp ( as it is by resize ), then we invent piles of extra pixels to filter together to fill in all the parts of the tile that will never be read.

Probably not high urgency, since the immediate use we expect for deep is upsampling. Eventually, we should probably try to fix it for both deep and flat ( while being careful with how perf sensitive the flat case is ).

Reply:
> The other thing that came up when playing with this is that performance of resizing to a very small result is incredibly bad.

The other part of this is that when the output image is only one tile, we lose out on multithreading. So for deep at least, it seems like part of the solution might be to do some threading internal to a tile? But I agree this isn't high urgency at present - definitely not something for this PR.